### DISCUSSION TITLE
XFRM interface default route exclude GW

### DISCUSSION THREAD

**Discussion by sangdrax8** [strongswan/strongswan#455] [Q&A] [answered]
I have been looking and can't find any settings to do this automatically.  I am converting my client to XFRM interfaces, and would like to route all traffic to the gateway.  However setting the default route down the interface, breaks the connection.

I am assuming I could start looking at updown scripts and set static routes for the current default route before changing default down the ipsec0 interface... but was wondering if I have missed an option to allow strong swan to automatically exclude the traffic it needs to keep the tunnel up.

I looked at the traffic selectors but I didn't see an exclude option.  Creating a traffic selector for all IP ranges except the server is probably possible but doesn't feel much cleaner than updown scripts.

**Comment by Thermi** [comment]
You need to make sure traffic to the remote peer does not go into the XFRM interface.
The XFRM interfaces (same for any other type of tunnel interface) process any packets you feed into them. There's no smart exemption.

**Comment by tobiasbrunner** [comment, marked as answer]
You could configure `charon.plugins.socket-default.fwmark` and then install your routes in a separate routing table for which you have a rule with `not fwmark <configured mark value>` (actually, if you also configure `charon.plugins.kernel-netlink.fwmark=!<mark value>` strongSwan will add such a rule for table 220 and you could install your routes there). That way the IKE packets won't be affected by your routes and continue to use the other tables (e.g. the default route in the main table).

**Reply by sangdrax8** [reply]
So I thought that looked promising, but as best I can tell the socket-default is setting the mark on IKE traffic, but my ESP (even if forced to UDP encap) doesn't get that mark and therefore still has the same issue of going down the tunnel.

However, since I am dealing with a single IP, I am able to get close enough by creating my own ip route table, and creating a rule that excludes my IPSec GW's address.  I believe this will be good enough for me, and now I can just worry about an up/down script for strongswan which adds or removes that rule which can default all other traffic down my ipsec0 xfrm interface.

Really feels like I am hacking this together, but hey I think I have a path forward that can at least work with as few additional scripts as possible.

**Reply by tobiasbrunner** [reply]
> So I thought that looked promising, but as best I can tell the socket-default is setting the mark on IKE traffic, but my ESP (even if forced to UDP encap) doesn't get that mark and therefore still has the same issue of going down the tunnel.

You could configure `set_mark_out` in swanctl.conf to the same value so ESP packets receive the same mark (requires a 4.19+ kernel. Or you mark ESP packets via firewall rule (not sure if you can do that early enough to affect the routing decision).

But yeah, you can also just install a route to your VPN gateway's IP address.

**Reply by Thermi** [reply]
set_mark_out and charon.plugins.socket-default.fwmark are the way to go

**Reply by sangdrax8** [reply]
That seems to do it!  I still used an updown script, but at least all I need to do there is create my XFRM interface, and then set my default in table 220.  That is simple enough that it shouldn't often have issues.  Thanks!
